frame/support/test/tests/storage_layers.rs

```diff
@@ -0,0 +1,242 @@
+			Ok(())
+		}));
+
+		assert_eq!(Value::<Runtime>::get(), 99);
+		assert_eq!(Map::<Runtime>::get(0), 99);
+	});
+}
+
+#[test]
+fn storage_layer_basic_rollback() {
+	TestExternalities::default().execute_with(|| {
+		assert_eq!(Value::<Runtime>::get(), 0);
+		assert_eq!(Map::<Runtime>::get(0), 0);
+
+		assert_noop!(
+			with_storage_layer(|| -> DispatchResult {
+				Value::<Runtime>::set(99);
+				Map::<Runtime>::insert(0, 99);
+				assert_eq!(Value::<Runtime>::get(), 99);
+				assert_eq!(Map::<Runtime>::get(0), 99);
+				Err("revert".into())
+			}),
+			"revert"
+		);
+
+		assert_eq!(Value::<Runtime>::get(), 0);
+		assert_eq!(Map::<Runtime>::get(0), 0);
+	});
+}
+
+#[test]
+fn storage_layer_rollback_then_commit() {
+	TestExternalities::default().execute_with(|| {
+		Value::<Runtime>::set(1);
+		Map::<Runtime>::insert(1, 1);
+
+		assert_ok!(with_storage_layer(|| -> DispatchResult {
+			Value::<Runtime>::set(2);
+			Map::<Runtime>::insert(1, 2);
+			Map::<Runtime>::insert(2, 2);
+
+			assert_noop!(
+				with_storage_layer(|| -> DispatchResult {
+					Value::<Runtime>::set(3);
+					Map::<Runtime>::insert(1, 3);
+					Map::<Runtime>::insert(2, 3);
+					Map::<Runtime>::insert(3, 3);
+
+					assert_eq!(Value::<Runtime>::get(), 3);
+					assert_eq!(Map::<Runtime>::get(1), 3);
+					assert_eq!(Map::<Runtime>::get(2), 3);
+					assert_eq!(Map::<Runtime>::get(3), 3);
+
+					Err("revert".into())
+				}),
+				"revert"
+			);
+
+			assert_eq!(Value::<Runtime>::get(), 2);
+			assert_eq!(Map::<Runtime>::get(1), 2);
+			assert_eq!(Map::<Runtime>::get(2), 2);
+			assert_eq!(Map::<Runtime>::get(3), 0);
+
+			Ok(())
+		}));
+
+		assert_eq!(Value::<Runtime>::get(), 2);
+		assert_eq!(Map::<Runtime>::get(1), 2);
+		assert_eq!(Map::<Runtime>::get(2), 2);
+		assert_eq!(Map::<Runtime>::get(3), 0);
+	});
+}
+
+#[test]
+fn storage_layer_commit_then_rollback() {
+	TestExternalities::default().execute_with(|| {
+		Value::<Runtime>::set(1);
+		Map::<Runtime>::insert(1, 1);
+
+		assert_noop!(
+			with_storage_layer(|| -> DispatchResult {
+				Value::<Runtime>::set(2);
+				Map::<Runtime>::insert(1, 2);
+				Map::<Runtime>::insert(2, 2);
+
+				assert_ok!(with_storage_layer(|| -> DispatchResult {
+					Value::<Runtime>::set(3);
+					Map::<Runtime>::insert(1, 3);
+					Map::<Runtime>::insert(2, 3);
+					Map::<Runtime>::insert(3, 3);
+
+					assert_eq!(Value::<Runtime>::get(), 3);
+					assert_eq!(Map::<Runtime>::get(1), 3);
+					assert_eq!(Map::<Runtime>::get(2), 3);
+					assert_eq!(Map::<Runtime>::get(3), 3);
+
+					Ok(())
+				}));
+
+				assert_eq!(Value::<Runtime>::get(), 3);
+				assert_eq!(Map::<Runtime>::get(1), 3);
+				assert_eq!(Map::<Runtime>::get(2), 3);
+				assert_eq!(Map::<Runtime>::get(3), 3);
+
+				Err("revert".into())
+			}),
+			"revert"
+		);
+
+		assert_eq!(Value::<Runtime>::get(), 1);
+		assert_eq!(Map::<Runtime>::get(1), 1);
+		assert_eq!(Map::<Runtime>::get(2), 0);
+		assert_eq!(Map::<Runtime>::get(3), 0);
+	});
+}
+
+#[test]
```

Comment:
would like to see one test where we execute a `call` that returns an error, and we make sure it is reverted.

This should prevent anyone from accidentally removing the 1 storage layer that is injected in the macro.

Reply:
That test is in the `assert_noop` below